static/js/protein/chart.ts

```diff
@@ -242,34 +258,117 @@ const LINK_STYLE = {
+  d3.select(`#${elementID}`).style(
+    "filter",
+    `brightness(${brightenPercentage})`
+  );
 }
 
-function unbrighten(): void {
-  // unbrighten object under cursor
-  d3.select(this).style("filter", "brightness(100%)");
+/**
+ * Select element by ID and reset brightness to 100%.
+ */
+function unbrighten(elementID: string): void {
+  d3.select(`#${elementID}`).style("filter", "brightness(100%)");
+}
+
+/**
+ * Given new html text, overwrite text currently in tooltip.
+ */
+function updateToolTipText(html: string): void {
+  TOOL_TIP.html(html);
 }
 
 /**
- * Gets the left and top coordinates of a rect element and positions the tooltip accordingly
+ * Get the left and top coordinates of a rect element and position the tooltip accordingly
  * @param left_position
  * @param top_position
  */
-function mousemove() {
-  TOOL_TIP.style("left", d3.event.pageX - 50 + "px").style(
+function updateToolTipPosition(): void {
+  TOOL_TIP.style("left", d3.event.pageX - 60 + "px").style(
     "top",
-    d3.event.pageY - 50 + "px"
+    d3.event.pageY - 60 + "px"
   );
 }
+
+/**
+ * Make tooltip visible.
+ */
+function showToolTip(): void {
+  TOOL_TIP.style("opacity", 1);
+}
+
 /**
- * Sets the opacity of the tooltip as zero, to hide it when the user hovers over other elements
+ * Make tooltip invisible.
  */
-function mouseout() {
+function hideToolTip(): void {
   TOOL_TIP.style("opacity", 0);
 }
+
+/**
+ * When mouse first enters element specified by given id, brighten it and update/display the global tooltip.
+ */
+function onMouseOver(
+  elementID: string,
+  toolTipText: string,
+  brightenPercentage: string = DEFAULT_BRIGHTEN_PERCENTAGE
+): void {
+  brighten(elementID, brightenPercentage);
+  updateToolTipText(toolTipText);
+  showToolTip();
+}
+
+/**
+ * Update position of global tooltip to track mouse.
+ */
+function onMouseMove(): void {
+  updateToolTipPosition();
+}
+
+/**
+ * When mouse leaves element specified by given id, reset its brightness and hide the global tooltip.
+ */
+function onMouseOut(elementID: string): void {
+  unbrighten(elementID);
+  hideToolTip();
+}
+
+/**
+ * On mouse hover, select hovered element and
+ *  1) highlight it
+ *  2) update the global tooltip
+ *  3) show the global tooltip.
+ *
+ * Unhighlight and hide the global tooltip when the mouse leaves.
+ */
+function handleMouseEvents(
+  selection: d3.Selection<Selectable, Datum, SVGGElement, unknown>,
+  idFunc: (index: number) => string,
+  toolTipFunc: (datum: Datum) => string,
+  brightenPercentage: string = DEFAULT_BRIGHTEN_PERCENTAGE
+): void {
+  selection
+    .on("mouseover", (d, i) => {
+      onMouseOver(idFunc(i), toolTipFunc(d), brightenPercentage);
+    })
+    .on("mousemove", onMouseMove)
+    .on("mouseout", (d, i) => onMouseOut(idFunc(i)));
+}
+
+/**
+ * Get function that takes an index and returns an ID containing the index and chart ID.
+ */
+function getBarIDFunc(chartID: string): (index: number) => string {
```

Comment:
I think instead of making this specific to bar, this function can just return `${chartID}-${index}` or `{chartID}-datum-${index}` or something general so it can be used for all the charts

Reply:
Great suggestion! This takes care of my manual `circleIDFunc` now too!

```const circleIDFunc = getElementIDFunc(id, 'circle');```